Run ObjCContractPass in Default Codegen Pipeline

NuriAmari opened the pull request:
Prior to this patch, when using -fthinlto-index= the ObjCARCContractPass isn't run prior to CodeGen, and instruction selection fails on IR containing arc intrinsics.

The pass would normally be added here: https://github.com/llvm/llvm-project/blob/c00e012bcf5da384a3e7339dc2e046779b339063/clang/lib/CodeGen/BackendUtil.cpp#L1352 Since we are using an index file, we return early above after `runThinLTOBackend`:
https://github.com/llvm/llvm-project/blob/c00e012bcf5da384a3e7339dc2e046779b339063/clang/lib/CodeGen/BackendUtil.cpp#L1333-L1339

Commit:
Run ObjCContractPass in Distributed Thin-LTO Pipeline

Prior to this patch, when using -fthinlto-index= the
ObjCARCContractPass isn't run prior to CodeGen, and
instruction selection fails on IR containing
arc intrinstics.

fhahn commented:
Shouldn't this be added to the LTO code generator? In `libLTO` (used by Apple's linker) it is added here `llvm/lib/LTO/ThinLTOCodeGenerator.cpp`

kyulee-com reviewed (approved):
I think `ObjCARCContractPass` is ideally suited for MachO rather than ELF. There might be a consideration to add it conditionally, however, this could be excessive. Like the (full)LTO pass -- https://github.com/llvm/llvm-project/blob/main/llvm/lib/LTO/LTOCodeGenerator.cpp#L141, I think this approach results in a no-op for ELF and it should be okay.

NuriAmari commented:
> Shouldn't this be added to the LTO code generator? In `libLTO` (used by Apple's linker) it is added here `llvm/lib/LTO/ThinLTOCodeGenerator.cpp`

Presumably because the pass is likely not useful unless targeting MachO, LLD does this via configuration hook: https://reviews.llvm.org/D94547. I'm already doing it unconditionally, so what do you think about adding the pass here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/LTO/LTOBackend.cpp#L417, and deleting this: https://github.com/llvm/llvm-project/blob/main/lld/MachO/LTO.cpp#L51-L53 ?

NuriAmari commented on clang/test/CodeGen/thinlto-distributed-objc-contract-pass.ll:
```diff
@@ -0,0 +1,19 @@
+target triple = "x86_64-apple-darwin"
+
+define void @use_arc(ptr %a, ptr %b) {
+  call void (...) @llvm.objc.clang.arc.use(i8* %a, i8* %b) nounwind
```
TODO: `i8*` -> `ptr`

Commit:
Address PR Feedback

- Adjust test
- Unconditionally include ARC pass in ThinLTO pipeline

NuriAmari commented:
> > Shouldn't this be added to the LTO code generator? In `libLTO` (used by Apple's linker) it is added here `llvm/lib/LTO/ThinLTOCodeGenerator.cpp`
> 
> Presumably because the pass is likely not useful unless targeting MachO, LLD does this via configuration hook: https://reviews.llvm.org/D94547. I'm already doing it unconditionally, so what do you think about adding the pass here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/LTO/LTOBackend.cpp#L417, and deleting this: https://github.com/llvm/llvm-project/blob/main/lld/MachO/LTO.cpp#L51-L53 ?

I've made this change

kyulee-com commented:
Because you now add it to `codegen` unconditionally, do you also need to delete https://github.com/llvm/llvm-project/blob/main/llvm/lib/LTO/LTOCodeGenerator.cpp#L141?

aeubanks commented:
it seems like this should just be in the default codegen pipeline? you'd need to change the pass to bail out early if there are no relevant intrinsics (by checking if the module contains the intrinsic declaration) to not affect compile times

cachemeifyoucan commented:
> it seems like this should just be in the default codegen pipeline? you'd need to change the pass to bail out early if there are no relevant intrinsics (by checking if the module contains the intrinsic declaration) to not affect compile times



> it seems like this should just be in the default codegen pipeline? you'd need to change the pass to bail out early if there are no relevant intrinsics (by checking if the module contains the intrinsic declaration) to not affect compile times



> it seems like this should just be in the default codegen pipeline? you'd need to change the pass to bail out early if there are no relevant intrinsics (by checking if the module contains the intrinsic declaration) to not affect compile times

Maybe that is a good idea since the check is already there and the check is quite cheap (just check the existence of ARC intrinsics).

I don't know about how lld is configured, but the concept of the change SGTM. The pass is missing probably because we don't link with distributed thinLTO that uses index-only option.

Commit:
PR Feedback #2

- Include Pass in default codegen pipeline
- Delete all other places the pass is added
- Bail early if no arc Intrinsics are found in the module

cachemeifyoucan commented:
> I couldn't find this check you're alluding to, so I added my own. If the check does exist, could you point me to it? If not, does mine look reasonable? Thanks.

I was answering from my memory, now I have to check. We have this function `ModuleHasARC` which is used in other ARC optimization passes. It used to be in arc-contract pass too but it seems to be dropped by https://reviews.llvm.org/D92808

It is a good idea to added it back. Maybe by reusing the same function after double check the context of the old patch. @ahatanaka might have idea if there is a reason to drop the guard.

Commit:
PR Feedback #3

- Replace custom predicate for arc intrinsics in a module with existing
  ModuleHasARC
- Remove unused include introduced by earlier commit

NuriAmari commented:
I've added the check back in, and reverted my custom check. There is too much going on in  https://reviews.llvm.org/D92808 for me to quickly understand everything, but I don't see an obvious reason the check needed to be removed. @ahatanaka Please let me know if I'm missing something.

Commit:
Remove unused include statement

Commit:
Fix pipeline tests to reflect changes

Commit:
Add 'llvm.objc.clang.arc.noop.use' to ModuleHasARC Predicate

This intrinsic was missing from the list the predicate checks a module
for in order to determine if ARC passes need to run. This results in
instruction selection failure if this instrinsic is left behind.

Add the intrinsic to the predicate to fix.

nikic commented:
Reverted due to major compile-time regressions for unoptimized builds, see https://llvm-compile-time-tracker.com/compare.php?from=910292c3ac2ebe43cdbc90223c6c9702128316db&to=8cc8e5d6c6ac9bfc888f3449f7e424678deae8c2&stat=instructions:u.

nikic commented on llvm/test/CodeGen/AArch64/O0-pipeline.ll:
```diff
@@ -31,6 +31,10 @@
 ; CHECK-NEXT:       Exception handling preparation
+; CHECK-NEXT:       Dominator Tree Construction
+; CHECK-NEXT:       Basic Alias Analysis (stateless AA impl)
+; CHECK-NEXT:       Function Alias Analysis Results
```
Presumably the extra analysis passes are the reason for the compile-time regressions. While you may be skipping the execution of the pass itself, you are not skipping the execution of the analyses.

aeubanks replied:
this is where having the new pass manager for codegen would be nice

NuriAmari replied:
I suspected this might be an issue. Is it possible to "lazily" request analysis results based on some logic in the pass itself? If not, maybe we can't enable this pass by default after all.
